python/cudf/cudf/_libxx/scalar.pyx

```diff
@@ -0,0 +1,218 @@
+            return _get_np_scalar_from_timestamp64(self.c_value)
+        else:
+            raise ValueError(
+                "Could not convert cudf::scalar to a Python value"
+            )
+
+    def __repr__(self):
+        if self.value is None:
+            return f"Scalar({self.value}, {self.dtype.__repr__()})"
+        else:
+            return f"Scalar({self.value.__repr__()})"
+
+    @staticmethod
+    cdef Scalar from_unique_ptr(unique_ptr[scalar] ptr):
+        """
+        Construct a Scalar object from a unique_ptr<cudf::scalar>.
+        """
+        cdef Scalar s = Scalar.__new__(Scalar)
+        s.c_value = move(ptr)
+        return s
+
+
+cdef _set_string_from_np_string(unique_ptr[scalar]& s, value, bool valid=True):
+    value = value.item() if valid else ""
+    s.reset(new string_scalar(value.encode(), valid))
+
+
+cdef _set_numeric_from_np_scalar(unique_ptr[scalar]& s,
+                                 value,
+                                 dtype,
+                                 bool valid=True):
+    value = value if valid else 0
+    if dtype == "int8":
+        s.reset(new numeric_scalar[int8_t](value, valid))
+    elif dtype == "int16":
+        s.reset(new numeric_scalar[int16_t](value, valid))
+    elif dtype == "int32":
+        s.reset(new numeric_scalar[int32_t](value, valid))
+    elif dtype == "int64":
+        s.reset(new numeric_scalar[int64_t](value, valid))
+    elif dtype == "float32":
+        s.reset(new numeric_scalar[float](value, valid))
+    elif dtype == "float64":
+        s.reset(new numeric_scalar[double](value, valid))
+    elif dtype == "bool":
+        s.reset(new numeric_scalar[bool8](<bool>value, valid))
+    else:
+        raise ValueError("dtype not supported: {}".format(dtype))
+
+
+cdef _set_datetime64_from_np_scalar(unique_ptr[scalar]& s,
+                                    value,
+                                    dtype,
+                                    bool valid=True):
+    value = value if valid else 0
+    if dtype == "datetime64[s]":
+        s.reset(
+            new timestamp_scalar[timestamp_s](<int64_t>np.int64(value), valid)
+        )
+    elif dtype == "datetime64[ms]":
+        s.reset(
+            new timestamp_scalar[timestamp_ms](<int64_t>np.int64(value), valid)
+        )
+    elif dtype == "datetime64[us]":
+        s.reset(
+            new timestamp_scalar[timestamp_us](<int64_t>np.int64(value), valid)
+        )
+    elif dtype == "datetime64[ns]":
+        s.reset(
+            new timestamp_scalar[timestamp_ns](<int64_t>np.int64(value), valid)
+        )
+    else:
+        raise ValueError("dtype not supported: {}".format(dtypne))
+
+
+cdef _get_py_string_from_string(unique_ptr[scalar]& s):
+    if not s.get()[0].is_valid():
+        return None
+    return (<string_scalar*>s.get())[0].to_string().decode()
+
+
+cdef _get_np_scalar_from_numeric(unique_ptr[scalar]& s):
+    if not s.get()[0].is_valid():
```

Comment:
We should be able to store the result of `s.get()[0]` once and then reuse it here and below.